transaction/std/src/test_utils.rs

```diff
@@ -201,3 +206,72 @@ pub fn get_transaction<RNG: RngCore + CryptoRng, FPR: FogPubkeyResolver + Clone>
 
     transaction_builder.build(rng)
 }
+
+/// Build simulated change memo with zero amount
+pub fn build_zero_value_change_memo(
+    builder: &mut impl MemoBuilder,
+) -> Result<MemoPayload, NewMemoError> {
+    build_change_memo_with_amount(builder, Amount::new(0, 0.into()))
+}
+
+/// Build simulated change memo with amount
+pub fn build_change_memo_with_amount(
+    builder: &mut impl MemoBuilder,
+    change_amount: Amount,
+) -> Result<MemoPayload, NewMemoError> {
+    // Create simulated context
+    let mut rng: StdRng = SeedableRng::from_seed([0u8; 32]);
+    let alice = AccountKey::random_with_fog(&mut rng);
+    let alice_address_book = ReservedDestination::from(&alice);
+    let change_tx_pubkey = RistrettoPublic::from_random(&mut rng);
+    let memo_context = MemoContext {
+        tx_public_key: &change_tx_pubkey,
+    };
+
+    //Build memo
+    builder.make_memo_for_change_output(change_amount, &alice_address_book, memo_context)
+}
+
+/// Utilities for decoding and verifying memo data
+pub struct MemoDecoder;
+
+impl MemoDecoder {
+    /// Decode a sender note from bytes
+    pub fn decode_sender_note(memo_data: &[u8; GiftCodeSenderMemo::MEMO_DATA_LEN]) -> &str {
+        let index = if let Some(terminator) = memo_data.iter().position(|b| b == &0u8) {
+            terminator
+        } else {
+            GiftCodeSenderMemo::MEMO_DATA_LEN
+        };
+
+        std::str::from_utf8(&memo_data[0..index]).unwrap()
+    }
+
+    /// Decode a funding note from bytes
+    pub fn decode_funding_note(memo_data: &[u8; GiftCodeFundingMemo::MEMO_DATA_LEN]) -> &str {
+        let index = if let Some(terminator) = memo_data
+            .iter()
+            .enumerate()
+            .position(|(i, b)| i >= GiftCodeFundingMemo::HASH_DATA_LEN && b == &0u8)
+        {
+            terminator
+        } else {
+            GiftCodeFundingMemo::MEMO_DATA_LEN
+        };
+
+        std::str::from_utf8(&memo_data[GiftCodeFundingMemo::HASH_DATA_LEN..index]).unwrap()
+    }
+
+    /// Get the first four bytes of the Blake2b hash of a gift code TxOut public
+    /// key
+    pub fn tx_out_public_key_short_hash(
```

Comment:
This probably belongs on the funding memo code, but is there any concern grabbing the first 4 bytes of a hash?
I don't know enough about the hash implementations, but could this result in the hashing logic ignoring the last n bytes of the input?  i.e. why do a 512bit hash on a value if only 32 bits are used, why not just do a 32 bit hash?
Again outside the scope of this review just a general question

Reply:
>  i.e. why do a 512bit hash on a value if only 32 bits are used, why not just do a 32 bit hash?

Because 32 bits would have a much larger possible set of collisions for different inputs. With a 512 bit output space, that set of possible colliding inputs is much larger. Using the first 4 bytes also increases the collision space, but not more than only using 32 bits. As far as why we used only 4 bytes of the 64 bytes of the hash in the MCIP - it's basically because we wanted to have as much space as possible for the text bytes.

It's a tradeoff we made :), but one that's unlikely to be attacked. If you do find some magic input that m

Reply:
> Using the first 4 bytes also increases the collision space, but not more than only using 32 bits.
I think this is better wording of what I was trying to ask, "is using the first 4 bytes the same collision space as using a dedicated 32 bit hash, or is it a larger collision space?"
I think for this use case it probably doesn't matter too much, just my lack of knowledge on how hash functions work under the hood.